dsa_test.go

```diff
@@ -0,0 +1,205 @@
+package openssl_test
+
+import (
+	"crypto/dsa"
+	"encoding/asn1"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/golang-fips/openssl/v2"
+	"github.com/golang-fips/openssl/v2/bbig"
+)
+
+type dsaSignature struct {
+	R, S *big.Int
+}
+
+func TestDSAGenerateParameters(t *testing.T) {
+	var tests = []struct {
+		L, N int
+	}{
+		{1024, 160},
+		{2048, 224},
+		{2048, 256},
+		{3072, 256},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%d-%d", test.L, test.N), func(t *testing.T) {
+			if openssl.FIPS() {
+				t.Skip("generating DSA parameters with L = 2048 is not supported in FIPS mode")
+			}
+			testGenerateDSAParameters(t, test.L, test.N)
+		})
+	}
+}
+
+func testGenerateDSAParameters(t *testing.T, L, N int) {
+	params, err := openssl.GenerateDSAParameters(L, N)
+	if err != nil {
+		t.Errorf("error generating parameters: %s", err)
+		return
+	}
+
+	P := bbig.Dec(params.P)
+	Q := bbig.Dec(params.Q)
+	G := bbig.Dec(params.G)
+
+	if P.BitLen() != L {
+		t.Errorf("params.BitLen got:%d want:%d", P.BitLen(), L)
+	}
+
+	if Q.BitLen() != N {
+		t.Errorf("q.BitLen got:%d want:%d", Q.BitLen(), L)
+	}
+
+	one := new(big.Int)
+	one.SetInt64(1)
+	pm1 := new(big.Int).Sub(P, one)
+	quo, rem := new(big.Int).DivMod(pm1, Q, new(big.Int))
+	if rem.Sign() != 0 {
+		t.Error("p-1 mod q != 0")
+	}
+	x := new(big.Int).Exp(G, quo, P)
+	if x.Cmp(one) == 0 {
+		t.Error("invalid generator")
+	}
+
+	priv, err := openssl.GenerateKeyDSA(params)
+	if err != nil {
+		t.Errorf("error generating key: %s", err)
+		return
+	}
+
+	testDSASignAndVerify(t, priv)
+}
+
+func testDSASignAndVerify(t *testing.T, priv *openssl.PrivateKeyDSA) {
```

Comment:
Should we test Sign separately from verify or would that never make sense?

Reply:
For the common cases it makes sense to exercise both in the same test, else we would have to hardcode the test vectors inputs, which is always more obscure. Note that we do hardcode inputs and tests just the signing or the verification function for some degenerated cases, see `TestDSANewPrivateKeyWithDegenerateKeys` and `TestDSANewPublicKeyWithBadPublicKey`.

At the end is a matter of taste. Here I'm following the idiosyncrasy of the Go standard library, which prefers tests covering multiple function calls (almost like integration tests) over small and atomic unit tests. IMO it makes sense to follow this approach here too, as most of the tests are inspired by tests in the standard library (for example, these tests are very similar to the ones in https://github.com/golang/go/blob/a9e6a96ac092cc5191d759488ee761e9a403ab8f/src/crypto/dsa/dsa_test.go).